File: starters/go/calculate.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type StationData struct {
	minimum float64
	maximum float64
	sum     float64
	count   int
}

func main() {
	inputFile := "measurements.txt"
	if len(os.Args) > 1 {
		inputFile = os.Args[1]
	}
	fmt.Fprintln(os.Stderr, "Reading records from", inputFile)

	file, err := os.Open(inputFile)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	start := time.Now()
	countLines := 0
	results, err := parseLines(file)
	if err != nil {
		panic(err)
	}
	for k, v := range results {
		countLines += v.count
		fmt.Printf("%s;%.1f;%.1f;%.1f\n", k, v.minimum, v.sum/float64(v.count), v.maximum)
	}
	duration := time.Now().Sub(start)
	fmt.Fprintf(os.Stderr, "Read %d measurements in %s\n", countLines, duration.Abs().Truncate(time.Millisecond))
}

func parseLines(file *os.File) (map[string]StationData, error) {
	results := make(map[string]StationData)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		split := strings.Split(line, ";")
		stationName := split[0]
		measurement, err := strconv.ParseFloat(split[1], 64)
		if err != nil {
			return nil, err
		}
		if entry, found := results[stationName]; found {
			results[stationName] = StationData{
				maximum: max(entry.maximum, measurement),
				minimum: min(entry.minimum, measurement),
				sum:     entry.sum + measurement,
				count:   entry.count + 1,
			}
		} else {
			results[stationName] = StationData{
				maximum: measurement,
				minimum: measurement,
				sum:     measurement,
				count:   1,
			}
		}
	}
	return results, scanner.Err()
}

File: starters/calculate.cpp
#include <chrono>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <sstream>
#include <string>
#include <unordered_map>
using namespace std;

struct StationData {
  float minMeasurement;
  float maxMeasurement;
  float sum;
  int count;
  StationData()
      : maxMeasurement(numeric_limits<float>::lowest()),
        minMeasurement(numeric_limits<float>::max()),
        sum(0),
        count(0) {}
};

void parseLine(unordered_map<string, StationData> *results, const string *line);

int main(int argc, char *argv[]) {
  string filename = "sample.txt";
  if (argc > 1) filename = argv[1];
  cout << "Reading records from " << filename << endl;

  ifstream file(filename);
  string line;

  if (!file.is_open()) {
    cerr << "Could not open file " << filename << endl;
    return 1;
  }

  auto start = chrono::high_resolution_clock::now();
  unordered_map<string, StationData> results;
  while (getline(file, line)) {
    parseLine(&results, &line);
  }
  file.close();
  auto end = chrono::high_resolution_clock::now();

  int countMeasurements(0);
  for (const auto &pair : results) {
    countMeasurements += pair.second.count;
    float mean = pair.second.sum / pair.second.count;
    cout << pair.first << ";" << fixed << setprecision(1)
         << pair.second.minMeasurement << ";" << mean << ";"
         << pair.second.maxMeasurement << endl;
  }

  chrono::duration<double> elapsed = end - start;
  double total_seconds = elapsed.count();
  int minutes = static_cast<int>(total_seconds / 60);
  double seconds = total_seconds - (minutes * 60);

  cerr << "Read " << countMeasurements << " measurements in " << minutes << "m"
       << fixed << setprecision(3) << seconds << "s" << endl;
  return 0;
}

void parseLine(unordered_map<string, StationData> *results,
               const string *line) {
  stringstream ss(*line);
  string stationName, measurementStr;
  getline(ss, stationName, ';');
  getline(ss, measurementStr);
  float measurement = stof(measurementStr);

  StationData current = (*results)[stationName];
  current.sum += measurement;
  current.minMeasurement = min(current.minMeasurement, measurement);
  current.maxMeasurement = max(current.maxMeasurement, measurement);
  current.count += 1;
  (*results)[stationName] = current;
}
